enterprise_gateway/services/sessions/kernelsessionmanager.py

```diff
@@ -279,21 +291,47 @@ def persistence_root_default(self):
 
-    def save_sessions(self):
+    def delete_sessions(self):
         if self.enable_persistence:
-            with open(self.kernel_session_file, 'w') as fp:
-                json.dump(KernelSessionManager.pre_save_transformation(self._sessions), fp)
-                fp.close()
+            for kernel_id in self.sessions_to_remove:
+                kernel_file_name = "".join([kernel_id, '.json'])
+                kernel_session_file_path = os.path.join(self._get_sessions_loc(), kernel_file_name)
+                if os.path.exists(kernel_session_file_path):
+                    os.remove(kernel_session_file_path)
+
+    def save_session(self, kernel_id):
+        if self.enable_persistence:
+            if kernel_id is not None:
+                kernel_file_name = "".join([kernel_id, '.json'])
+                kernel_session_file_path = os.path.join(self._get_sessions_loc(), kernel_file_name)
+                temp_session = dict()
+                temp_session[kernel_id] = self._sessions[kernel_id]
+                with open(kernel_session_file_path, 'w') as fp:
+                    json.dump(KernelSessionManager.pre_save_transformation(temp_session), fp)
+                    fp.close()
 
     def load_sessions(self):
         if self.enable_persistence:
-            if os.path.exists(self.kernel_session_file):
-                self.log.debug("Loading saved sessions from {}".format(self.kernel_session_file))
-                with open(self.kernel_session_file) as fp:
-                    self._sessions = KernelSessionManager.post_load_transformation(json.load(fp))
+            kernel_session_files = [json_files for json_files in os.listdir(self._get_sessions_loc()) if
+                                    json_files.endswith('.json')]
+            for kernel_session_file in kernel_session_files:
+                kernel_session_file_path = os.path.join(self._get_sessions_loc(), kernel_session_file)
+                if os.path.exists(kernel_session_file_path):
+                    self.log.debug("Loading saved session(s) from {}".format(kernel_session_file_path))
+                    with open(kernel_session_file_path) as fp:
+                        self._sessions.update(KernelSessionManager.post_load_transformation(json.load(fp)))
+                        fp.close()
+
+    def load_session(self, kernel_id):
+        if kernel_id is not None:
+            kernel_file_name = "".join([kernel_id, '.json'])
+            kernel_session_file_path = os.path.join(self._get_sessions_loc(), kernel_file_name)
+            if os.path.exists(kernel_session_file_path):
+                self.log.debug("Loading saved session(s) from {}".format(kernel_session_file_path))
+                with open(kernel_session_file_path) as fp:
+                    self._sessions.update(KernelSessionManager.post_load_transformation(json.load(fp)))
                     fp.close()
```

Comment:
The previous 6 lines are duplciated in `load_sessions`.  Let's create a private method, `_load_session_from_file(file_name)` that is essentially this...
```python
        def _load_session_from_file(self, file_name):
            kernel_session_file_path = os.path.join(self._get_sessions_loc(), file_name)
            if os.path.exists(kernel_session_file_path):
                self.log.debug("Loading saved session(s) from {}".format(kernel_session_file_path))
                with open(kernel_session_file_path) as fp:
                    self._sessions.update(KernelSessionManager.post_load_transformation(json.load(fp)))
                    fp.close()
```
and replace the duplicated lines.

Reply:
Done.